power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -714,52 +766,174 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
+                std::vector<bool> id_na{};
+                for (const auto& obj : span) {
+                    if constexpr (requires { obj.id; }) {
+                        id_na.emplace_back(is_nan(obj.id));
+                    }
+                }
+                ids_na.emplace_back(std::move(id_na));
+            }
+            return ids_na;
+        };
 
-        auto const process_buffer_span = []<typename CT>(auto const& all_spans) -> bool {
+        auto process_buffer_span = [check_ids_na]<typename CT>(auto const& all_spans, UpdateCompProperties& result) {
             // Remember the first batch size, then loop over the remaining batches and check if they are of the same
             // length
-            auto const elements_per_scenario = static_cast<Idx>(all_spans.front().size());
-            bool const uniform_batch = std::ranges::all_of(all_spans, [elements_per_scenario](auto const& span) {
-                return static_cast<Idx>(span.size()) == elements_per_scenario;
+            std::vector<std::vector<bool>> const ids_na = check_ids_na(all_spans);
+            result.has_id = !std::ranges::all_of(ids_na, [](const std::vector<bool>& vec) { return vec.empty(); });
+            result.ids_all_na = std::ranges::all_of(ids_na, [](const std::vector<bool>& vec) {
+                return std::ranges::all_of(vec, [](bool const& obj) { return obj; });
             });
-            if (!uniform_batch) {
-                return false;
-            }
-            if (elements_per_scenario == 0) {
-                return true;
-            }
-            // Remember the begin iterator of the first scenario, then loop over the remaining scenarios and check the
-            // ids
+            result.ids_part_na = std::ranges::any_of(ids_na, [](const std::vector<bool>& vec) {
+                return std::ranges::any_of(vec, [](bool const& obj) { return obj; }) &&
+                       std::ranges::any_of(vec, [](bool const& obj) { return !obj; });
+            });
+            result.uniform = std::ranges::all_of(
+                all_spans, [n_elements = static_cast<Idx>(all_spans.front().size())](auto const& span) {
+                    return static_cast<Idx>(std::size(span)) == n_elements;
+                });
+            // Remember the begin iterator of the first scenario, then loop over the remaining scenarios and check
+            // the ids
             auto const first_span = all_spans[0];
             // check the subsequent scenarios
             // only return true if all scenarios match the ids of the first batch
-            return std::all_of(all_spans.cbegin() + 1, all_spans.cend(), [&first_span](auto const& current_span) {
-                return std::ranges::equal(
-                    current_span, first_span,
-                    [](UpdateType<CT> const& obj, UpdateType<CT> const& first) { return obj.id == first.id; });
-            });
+            result.ids_match =
```

Comment:
this ties into the discussion on terminology consistency:

```md
* **IDs match:** whether the dataset modifies the same components across its scenarios
  * uniform: all scenarios modify the exact same components in the exact same order across all scenarios
  * non-uniform: not all scenarios affect the same components in the same order
* **dataset uniformity [code]:** whether the dataset has the content size across its scenarios
  * uniform: all scenarios modify the exact same amount of components **regardless of the way it is represented**
  * non-uniform: not all scenarios affect the same amount of components
* **dataset uniformity [docs]:** whether the dataset has the same shape across its scenarios
  * uniform: all scenarios modify the same amount of components **in a dense representation**
    equivalent to dense data (!!!)
  * non-uniform: not all scenarios affect an equal amount of components
    an indptr dictates where a scenario starts and ends
    equivalent to sparse data (!!!)
```

let's discuss tomorrow. in the meantime, I will review as-is

Reply:
Resolved in #818